models/Asset.php

```diff
@@ -1540,68 +1540,88 @@ public function getMetadata($name = null, $language = null, $strictMatch = false
         $this->dispatchEvent($preEvent, AssetEvents::PRE_GET_METADATA);
         $this->metadata = $preEvent->getArgument('metadata');
 
-        $convert = function ($metaData) {
-            $loader = Pimcore::getContainer()->get('pimcore.implementation_loader.asset.metadata.data');
-            $transformedData = $metaData['data'];
+        if ($name) {
+            return $this->getMetadataByName($name, $language, $strictMatch, $raw);
+        }
 
-            try {
-                /** @var Data $instance */
-                $instance = $loader->build($metaData['type']);
-                $transformedData = $instance->transformGetterData($metaData['data'], $metaData);
-            } catch (UnsupportedException $e) {
-            }
+        $metaData = $this->getObjectVar('metadata');
+        $result = [];
+        $metaDataWithLanguage = [];
+        if (is_array($metaData)) {
+            foreach ($metaData as $md) {
+                $md = (array)$md;
 
-            return $transformedData;
-        };
+                if ((empty($md['language']) && !$strictMatch) || ($language == $md['language']) || !$language) {
+                    if (!$raw) {
+                        $md['data'] = $this->transformMetadata($md);
+                    }
+                    $result[] = $md;
+                }
 
-        if ($name) {
-            if ($language === null) {
-                $language = Pimcore::getContainer()->get(LocaleServiceInterface::class)->findLocale();
+                if(!empty($md['language'])) {
+                    $metaDataWithLanguage[$md['language']][$md['name']] = $md;
+                }
             }
+        }
 
-            $data = null;
-            foreach ($this->metadata as $md) {
-                if ($md['name'] == $name) {
-                    if ($language == $md['language']) {
-                        if ($raw) {
-                            return $md;
-                        }
+        foreach($result as $key => &$item) {
+            if ($language && !$strictMatch) {
+                $lang = $item['language'];
 
-                        return $convert($md);
-                    }
-                    if (empty($md['language']) && !$strictMatch) {
-                        if ($raw) {
-                            return $md;
-                        }
-                        $data = $md;
+                if (!$lang && isset($metaDataWithLanguage[$language]) && isset($metaDataWithLanguage[$language][$item['name']])) {
```

Comment:
```suggestion
                if (!$item['language'] && isset($metaDataWithLanguage[$language][$item['name']])) {
```
and remove the assignment `$lang = $item['language'];`, since `$lang` not used elsewhere beside in this IF block.
In case it is better to keep it there, probably would rename `$lang` to something more descriptive.

The `isset` of the deeper array key would be enough